### ISSUE TITLE
roachtest: acceptance/version-upgrade is flaky

### ISSUE THREAD

**Issue by adityamaru** [cockroachdb/cockroach#87104]
In the past few days the acceptance/version-upgrade roachtest has been failing in various ways, some error modes are:

```
pq: failed to run backup: exporting 112 ranges: unable to dial n2: breaker open

dial tcp 127.0.0.1:26259: connect: connection refused

pq: version mismatch in flow request: 65; this node accepts 69 through 69
```

The last one is the most common failure mode at the moment where the test fails at this step - https://github.com/cockroachdb/cockroach/blob/b87e290a866b022efb81ed770fcf5716bbb9d142/pkg/cmd/roachtest/tests/versionupgrade.go#L149 which is when node 1 is running the current binary version, while the other nodes are on the predecessor binary versions.

Build examples:
https://teamcity.cockroachdb.com/viewLog.html?buildId=6278345&tab=buildResultsDiv&buildTypeId=Cockroach_Ci_Tests_LocalRoachtest
https://teamcity.cockroachdb.com/viewLog.html?buildId=6278433&buildTypeId=Cockroach_BazelEssentialCi

Jira issue: CRDB-19153

**Comment by tbg** [comment]
Marking as release-blocker to reflect the gravity of this flake - afaict it's likely a problem that would be encountered by customers' workloads while upgrading to 22.2.

I suggest someone from SQL queries to own this. @yuzefovich can you think of someone appropriate and facilitate the assignment? Thank you!

**Comment by renatolabs** [comment]
FWIW, I remember seeing the second error message when I was trying to reduce the flakiness of this test about a month ago (#84382), so I don't think it's new. However, it was a fairly rare occurrence, and maybe it's become more frequent since then.

**Comment by msirek** [comment]
Action item here may be to do a bisect.

**Comment by yuzefovich** [comment]
I believe I identified the root cause in #87154 (it's a test issue, not an actual bug), so removing the release blocker label.

**Comment by yuzefovich** [comment]
I saw it [flake](https://teamcity.cockroachdb.com/viewLog.html?buildId=6314258&buildTypeId=Cockroach_UnitTests) on one of my PRs with a different error at a different time in the test. I'll need to take another look.

**Comment by yuzefovich** [comment]
So here is what's happening in this [flake](https://teamcity.cockroachdb.com/viewLog.html?buildId=6314258&buildTypeId=Cockroach_UnitTests):
- n4 has just been re-upgraded from 22.1.6 to current
- n4 is the gateway for a `SELECT` query of `Object Access` feature
- n4 thinks n1 is the leaseholder for the relevant range, so n4 issues `SetupFlow` RPC to n1
- n1 receives that request and needs to perform `FlowStream` RPC to stream data back to n4.
- n1 is able to get a connection (because we [ignore](https://github.com/cockroachdb/cockroach/blob/809458bf850096c91894c361eb75ccee05212408/pkg/sql/execinfra/outboxbase.go#L47) the breaker), but then n1 fails to perform `FlowStream` RPC because of the breaker
```
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109  â€¹[core]â€ºâ€¹[Channel #22 SubChannel #23] grpc: addrConn.createTransport failed to connect to {â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "Addr": "127.0.0.1:26263",â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "ServerName": "127.0.0.1:26263",â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "Attributes": null,â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "BalancerAttributes": null,â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "Type": 0,â€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹  "Metadata": nullâ€º
W220901 00:36:29.692490 2795 google.golang.org/grpc/grpclog/component.go:41 â‹® [-] 109 +â€¹}. Err: connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º
W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110  Outbox FlowStream connection error, distributed query will fail: â€¹rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º
W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110 +(1) â€¹rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º
```
To me this looks like a dup of #44101.

I'm not sure what to do here though. I don't have much context of how we would go about fixing #44101. Couple of options for fixing this flake in particular:
- disallow distributing queries for a brief period of time when the recently started node is the gateway for the query. This will be similar in spirit to #44102.
- allow retrying / skipping the query if it hits "no inbound stream connection" error in the version upgrade test. We previously were already doing that until #52624 "solved" the problem. Seems like the recent change to stop the cockroach nodes gracefully increased the likelihood of flakes.

Curious what others think, especially @tbg on the feasibility of addressing #44101 for good.

**Comment by tbg** [comment]
> because of the breaker

I don't see the breaker error in the output you pasted. Rather, this is the `onlyOnceDialer`:

https://github.com/cockroachdb/cockroach/blob/34de5fb3a360bddc5e87d6aa842b8eec56b04ce9/pkg/rpc/context.go#L1548-L1558

meaning that a previous attempt to dial failed, and legitimately failed (i.e. wasn't stopped by the breaker, as this wouldn't waste the onlyOnceDialer). Could you pull up a bit more of the log to see if you can find the true reason n1 couldn't talk to n4?

**Comment by tbg** [comment]
re: "real" fix, see https://github.com/cockroachdb/cockroach/issues/44101#issuecomment-1236963852

**Comment by yuzefovich** [comment]
I copied the logs from [here](https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests/6314258:id/roachtest/acceptance/version-upgrade/run_1/artifacts.zip!/logs/1.unredacted/cockroach.c279cfdd62ad.roach.2022-09-01T00_36_10Z.009990.log). Do you mean getting a more verbose logging output that what is printed by default?

**Comment by tbg** [comment]
The logging just doesn't corroborate the scenario you've outlined, you say

> n1 is able to get a connection (because we [ignore](https://github.com/cockroachdb/cockroach/blob/809458bf850096c91894c361eb75ccee05212408/pkg/sql/execinfra/outboxbase.go#L47) the breaker), **but then n1 fails to perform FlowStream RPC because of the breaker**

That last part doesn't seem true - it looks more like the connection it pulled from the node dialer here

https://github.com/cockroachdb/cockroach/blob/711e22877089164172e7a8022ff0662bb9fc0daa/pkg/sql/execinfra/outboxbase.go#L32-L53

is somehow unhealthy? Is it possible that the DistSQL request somehow straddles the restart and that n4 legit was down (or hadn't fully restarted yet) when that query was run? The reason I suspect this is because there's lots of code that you're hitting that tries to establish this connection as healthy, 

https://github.com/cockroachdb/cockroach/blob/50c1196e50041bbe5d5584a4ad7f221613902213/pkg/rpc/nodedialer/nodedialer.go#L195-L227

**Comment by yuzefovich** [comment]
> Is it possible that the DistSQL request somehow straddles the restart

That doesn't seem possible because the query is issued only after n4 is restarted.

> n4 hadn't fully restarted yet

That seems plausible.

----

Here are the things that I'm confident in:
- n4 has just been upgraded
- n4 is the gateway for "Object Access" query and performs `SetupFlow` RPC against n1, which succeeds
- n1 serves `SetupFlow` RPC and creates an outbox
- that outbox is able to get a connection via `GetConnForOutbox`, but then `FlowStream` RPC against n4 fails with
```
W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110  Outbox FlowStream connection error, distributed query will fail: â€¹rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º
W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110 +(1) â€¹rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º
W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110 +Error types: (1) *status.Error
```
- n4 keeps on waiting for n1 to dial back in for 10 seconds, after which it times out the query with
```
E220901 00:36:39.693231 2632 sql/flowinfra/flow_registry.go:336 â‹® [n4,client=127.0.0.1:37748,user=root,fâ€¹ab65bb47â€º] 148  flow id:â€¹ab65bb47-63ac-4f0f-85e0-0a0a088503c0â€º : 1 inbound streams timed out after 10s; propagated error throughout flow
```

----

Let's take a closer look at the logs of n4 after the restart.
```
I220901 00:36:29.645727 132 1@server/server_sql.go:1415 â‹® [n4] 52  serving sql connections
...
I220901 00:36:29.648057 1012 upgrade/upgrademanager/manager.go:115 â‹® [n4,intExec=â€¹set-versionâ€º,migration-mgr] 54  migrating cluster from 22.1 to 1000022.1-68 (stepping through [1000022.1-2 1000022.1-4 1000022.1-6 1000022.1-8 1000022.1-10 1000022.1-12 1000022.1-14 1000022.1-16 1000022.1-18 1000022.1-20 1000022.1-22 1000022.1-24 1000022.1-26 1000022.1-28 1000022.1-30 1000022.1-32 1000022.1-34 1000022.1-36 1000022.1-38 1000022.1-40 1000022.1-42 1000022.1-44 1000022.1-46 1000022.1-48 1000022.1-50 1000022.1-52 1000022.1-54 1000022.1-56 1000022.1-58 1000022.1-60 1000022.1-62 1000022.1-64 1000022.1-66 1000022.1-68])
I220901 00:36:29.649797 1012 upgrade/upgradecluster/cluster.go:118 â‹® [n4,intExec=â€¹set-versionâ€º,migration-mgr] 55  executing validate-cluster-version=1000022.1-68 on nodes n{1,2,3,4}
I220901 00:36:29.695749 1012 upgrade/upgrademanager/manager.go:135 â‹® [n4,intExec=â€¹set-versionâ€º,migration-mgr] 56  stepping through 1000022.1-2
I220901 00:36:29.725805 1188 jobs/adopt.go:243 â‹® [n4,intExec=â€¹set-versionâ€º,migration-mgr] 57  job 792730631195164676: resuming execution
I220901 00:36:29.741825 1190 jobs/registry.go:1206 â‹® [n4] 58  MIGRATION job 792730631195164676: stepping through state running with error: <nil>
I220901 00:36:29.763583 1190 jobs/registry.go:1206 â‹® [n4] 59  MIGRATION job 792730631195164676: stepping through state succeeded with error: <nil>
...
I220901 00:36:29.990979 1474 jobs/registry.go:1206 â‹® [n4] 99  MIGRATION job 792730632062795780: stepping through state running with error: <nil>
```
If we can rely on the clocks of n1 and n4 being in sync, then we can see that n4 is still running through the upgrade migrations at the time when n1 tries to perform `FlowStream` RPC which seems to corroborate the theory that n4 wasn't fully "up" yet. Do you know whether running migrations on n4 would somehow prevent other nodes to dial into it? Are we starting to serve sql connections too early (i.e. should we wait for migrations to complete)?

**Comment by tbg** [comment]
`cockroach start` (as deployed by roachprod) will return when it hits the `sdnotify` line at the end of this method:

https://github.com/cockroachdb/cockroach/blob/1af66356e77f516ede7cd56a3d7946873a590ef7/pkg/cli/start.go#L489-L548

Since n4 is *restarted*, the relevant line is this:

https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/server/server.go#L1275

which is invoked at the top of the diff here:

https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/server/server.go#L1403-L1495

and note the bottom of the diff which sets grpc to "operational" (meaning it'll stop refusing incoming requests).

The listener is opened a few pages before this, so a dial to n4 should have succeeded (i.e. no conn refused or the like); before the `operational` line, RPCs would have been refused, but we are getting a different error which indicates that an attempt to dial failed:

> W220901 00:36:29.692553 2826 sql/colflow/colrpc/outbox.go:194 â‹® [n1,fâ€¹ab65bb47â€º,streamID=â€¹0â€º] 110  Outbox FlowStream connection error, distributed query will fail: â€¹rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing cannot reuse client connection"â€º

Unfortunately, the error we really want is the one "before" that; this error here only tells us that a previous dial failed. Why did it fail? With what? That is unclear.

**Comment by tbg** [comment]
@yuzefovich I made a separate issue for this problem: https://github.com/cockroachdb/cockroach/issues/87634

For now, let's introduce a 4s sleep after each node restart, that should reliably paper over it. Not great, but I don't think this is a new problem - I think we're seeing it now because we are now draining the nodes and so there is no range unavailability after downtime, which probably papered over it very reliably. Would you be able to send that PR, Yahor, and close this issue out if it passes a couple of runs?

**Comment by yuzefovich** [comment]
Thanks Tobi! I'll send a patch.

**Comment by tbg** [comment]
Closing since it's been passing for close to a week now. If it fails again, better to open a new issue.
